src/Confluent.Kafka.Avro/GenericDeserializerImpl.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright 2016-2018 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using Avro.IO;
+using Avro.Generic;
+using Confluent.SchemaRegistry;
+
+
+namespace Confluent.Kafka.Serialization
+{
+    internal class GenericDeserializerImpl : IAvroDeserializerImpl<GenericRecord>
+    {
+        /// <remarks>
+        ///     A datum reader cache (one corresponding to each write schema that's been seen)
+        ///     is maintained so that they only need to be constructed once.
+        /// </remarks>
+        private readonly Dictionary<int, DatumReader<GenericRecord>> datumReaderBySchemaId 
+            = new Dictionary<int, DatumReader<GenericRecord>>();
+
+        private object datumReaderLock = new object();
+
+        private ISchemaRegistryClient schemaRegistryClient;
+
+        public GenericDeserializerImpl(ISchemaRegistryClient schemaRegistryClient)
+        {
+            this.schemaRegistryClient = schemaRegistryClient;
+        }
+
+        public GenericRecord Deserialize(string topic, byte[] array)
+        {
+            // Note: topic is not necessary for deserialization (or knowing if it's a key 
+            // or value) only the schema id is needed.
+
+            using (var stream = new MemoryStream(array))
+            using (var reader = new BinaryReader(stream))
+            {
+                var magicByte = reader.ReadByte();
+                if (magicByte != Constants.MagicByte)
+                {
+                    // may change in the future.
+                    throw new InvalidDataException($"magic byte should be 0, not {magicByte}");
+                }
+                var writerIdBigEndian = reader.ReadInt32();
+                var writerId = IPAddress.NetworkToHostOrder(writerIdBigEndian);
+
+                DatumReader<GenericRecord> datumReader = null;
+
+                lock (datumReaderLock)
+                {
+                    datumReaderBySchemaId.TryGetValue(writerId, out datumReader);
+                    if (datumReader == null)
+                    {
+                        // TODO: If any of this cache fills up, this is probably an
```

Comment:
yes, this should really be an LRU

Reply:
more important though is the behavior if it does fill up, since adding an LRU will be just adding an optimization for a situation that generally shouldn't occur in practice.